proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy

```diff
@@ -304,30 +329,91 @@ lemma do_reply_transfer_arch_tcb_at[wp]:
   (wp: crunch_wps preemption_point_inv unless_wp mapME_x_wp'
    simp: crunch_simps)
 
 crunch invoke_untyped
-  for cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
-  and active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
+  for active_cur_vcpu_of[wp]: "\<lambda>s. P (active_cur_vcpu_of s)"
   (wp: crunch_wps mapME_x_inv_wp preemption_point_inv
    simp: crunch_simps mapM_x_def[symmetric] active_cur_vcpu_of_def)
 
+\<comment> \<open>The following @{term etcb_at} lemmas extend earlier ones that don't have the Qs. These have
+   extra preconditions however, so we keep the other lemmas around.\<close>
+lemma delete_objects_etcb_at':
+  "\<lbrace>\<lambda>s. Q (etcb_at P t s) \<and> (t \<notin> {ptr..ptr + 2 ^ bits - 1})\<rbrace>
+   delete_objects ptr bits
+   \<lbrace>\<lambda>_ s. Q (etcb_at P t s)\<rbrace>"
+  apply (simp add: delete_objects_def)
+  apply (wpsimp simp: detype_def)
+  apply (auto simp: detype_def etcbs_of'_def etcb_at'_def)
+  done
+
+lemma reset_untyped_cap_etcb_at':
+  "\<lbrace>\<lambda>s. Q (etcb_at P t s) \<and> invs s \<and> cte_wp_at (\<lambda>cp. t \<notin> cap_range cp \<and> is_untyped_cap cp) slot s\<rbrace>
+   reset_untyped_cap slot
+   \<lbrace>\<lambda>_ s. Q (etcb_at P t s)\<rbrace>"
+  apply (simp add: reset_untyped_cap_def)
+   apply (wpsimp wp: mapME_x_inv_wp preemption_point_inv get_cap_wp delete_objects_etcb_at')
+  apply (auto simp: cte_wp_at_caps_of_state bits_of_def is_cap_simps)
+  done
+
+lemma invoke_untyped_etcb_at':
+  "\<lbrace>\<lambda>s. Q (etcb_at P t s) \<and> invs s \<and> st_tcb_at (Not o inactive and Not \<circ> idle) t s \<and> ct_active s
+        \<and> valid_untyped_inv ui s\<rbrace>
+   invoke_untyped ui
+   \<lbrace>\<lambda>_ s. st_tcb_at (Not o inactive) t s \<longrightarrow> Q (etcb_at P t s)\<rbrace>"
+  apply (cases ui)
+  apply (simp add: mapM_x_def[symmetric] invoke_untyped_def)
+  apply (wpsimp wp: mapM_x_wp'
+                    create_cap_no_pred_tcb_at typ_at_pred_tcb_at_lift
+                    hoare_convert_imp[OF create_cap_no_pred_tcb_at]
+                    hoare_convert_imp[OF _ init_arch_objects_etcbs_of]
+                    hoare_drop_impE_E reset_untyped_cap_etcb_at'[where Q=Q])
+  apply (cases ui, clarsimp)
+  apply (frule(1) st_tcb_ex_cap[OF _ invs_iflive])
+   apply (clarsimp split: Structures_A.thread_state.splits)
+  apply (drule ex_nonz_cap_to_overlap,
+         ((simp add: cte_wp_at_caps_of_state descendants_range_def2 empty_descendants_range_in)+))
+  done
+
+lemma invoke_untyped_etcb_at:
+  "\<lbrace>\<lambda>s. Q (etcb_at P t s) \<and> invs s \<and> st_tcb_at (Not o inactive and Not \<circ> idle) t s \<and> ct_active s \<and> valid_untyped_inv ui s\<rbrace>
+   invoke_untyped ui
+   \<lbrace>\<lambda>_ s. Q (etcb_at P t s)\<rbrace>"
+  apply (rule hoare_post_imp[where Q'="\<lambda>_ s. st_tcb_at (Not \<circ> inactive) t s \<and> (st_tcb_at (Not \<circ> inactive) t s \<longrightarrow> Q (etcb_at P t s))"])
```

Comment:
feels like I commented on this before, but there the `invoke_untyped_etcb_at` got a `'` only, instead of appearing as an addition... bit confused what happened here